// node-graylog/lib/gelfMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var localIP_1 = require("./localIP");
var _ = require("lodash");
var pkg = require('../package.json');
var GelfMessage = /** @class */ (function () {
    function GelfMessage() {
        this.version = pkg.version;
        this._facility = 'node.js';
        this.host = localIP_1.default;
        this.timestamp = (new Date().getTime() / 1000);
    }
    GelfMessage.prototype.setApp = function (app) {
        this.app = app;
        return this;
    };
    GelfMessage.prototype.setHost = function (host) {
        if (host) {
            this.host = host;
        }
        return this;
    };
    GelfMessage.prototype.setLevel = function (level) {
        this.level = level;
        return this;
    };
    GelfMessage.prototype.setFile = function (file) {
        this.file = file;
        return this;
    };
    GelfMessage.prototype.setLine = function (line) {
        if (line) {
            this.line = line;
        }
        return this;
    };
    GelfMessage.prototype.setLogger = function (logger) {
        this.logger = logger;
        return this;
    };
    GelfMessage.prototype.setMarker = function (marker) {
        if (marker) {
            this.marker = marker;
        }
        return this;
    };
    GelfMessage.prototype.setThread = function (thread) {
        if (thread) {
            this.thread = thread;
        }
        return this;
    };
    GelfMessage.prototype.setRequestUri = function (requestUri) {
        if (requestUri) {
            this.requestUri = requestUri;
        }
        return this;
    };
    GelfMessage.prototype.setShortMessage = function (shortMessage) {
        this.shortMessage = shortMessage;
        return this;
    };
    GelfMessage.prototype.setFullMessage = function (fullMessage) {
        this.fullMessage = fullMessage;
        return this;
    };
    GelfMessage.prototype.setEnv = function (env) {
        this.env = env;
        return this;
    };
    GelfMessage.prototype.setCluster = function (cluster) {
        this.cluster = cluster;
        return this;
    };
    GelfMessage.prototype.toString = function () {
        return JSON.stringify(_.pickBy({
            version: this.version,
            app: this.app,
            host: this.host,
            file: this.file,
            logger: this.logger,
            short_message: this.shortMessage,
            full_message: this.fullMessage,
            request_uri: this.requestUri,
            marker: this.marker,
            thread: this.thread,
            line: this.line,
            _facility: this._facility,
            timestamp: this.timestamp,
            level: this.level,
            env: this.env,
            cluster: this.cluster
        }, function (o) { return o !== undefined && o !== null && o !== ''; }));
    };
    return GelfMessage;
}());
exports.default = GelfMessage;
