// node-graylog/lib/gelfMessageFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var gelfMessage_1 = require("./gelfMessage");
exports.LEVEL_MAPPING = {
    EMERGENCY: 0,
    ALERT: 1,
    CRITICAL: 2,
    ERROR: 3,
    WARN: 4,
    NOTICE: 5,
    INFO: 6,
    DEBUG: 7
};
var GelfMessageFactory = /** @class */ (function () {
    function GelfMessageFactory(_a) {
        var file = _a.file, app = _a.app, logger = _a.logger, marker = _a.marker, thread = _a.thread, requestUri = _a.requestUri, host = _a.host, env = _a.env, cluster = _a.cluster;
        this.level = exports.LEVEL_MAPPING.INFO;
        this.file = file;
        this.app = app;
        this.logger = logger;
        this.marker = marker;
        this.thread = thread;
        this.requestUri = requestUri;
        this.host = host;
        this.env = env;
        this.cluster = cluster;
    }
    GelfMessageFactory.prototype.createGelfMessage = function (msg, level) {
        return new gelfMessage_1.default()
            .setFile(this.file)
            .setLevel(level !== undefined ? exports.LEVEL_MAPPING[level.toUpperCase()] : this.level)
            .setApp(this.app)
            .setHost(this.host)
            .setLogger(this.logger)
            .setMarker(this.marker)
            .setThread(this.thread)
            .setEnv(this.env)
            .setCluster(this.cluster)
            .setRequestUri(this.requestUri)
            .setShortMessage((msg.length > 128) ? msg.substr(0, 1420 * 128) : msg)
            .toString();
    };
    return GelfMessageFactory;
}());
exports.default = GelfMessageFactory;
