// node-graylog/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var gelfUDPBalanceSender_1 = require("./gelfUDPBalanceSender");
__export(require("./gelfUDPBalanceSender"));
var Logger = /** @class */ (function () {
    function Logger(options) {
        var _this = this;
        this.senderHooks = {};
        this.sender = new gelfUDPBalanceSender_1.default(options);
        if (Array.isArray(options.hooks) && options.hooks.length) {
            options.hooks.forEach(function (h) {
                var _a;
                var logType = h.type;
                var hooks = h.hooks;
                if (!_this.senderHooks[logType]) {
                    _this.senderHooks[logType] = [];
                }
                (_a = _this.senderHooks[logType]).push.apply(_a, hooks);
            });
        }
    }
    Logger.prototype.log = function (msg, level) {
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        var _a;
        (_a = this.sender).sendMessage.apply(_a, [msg, level || 'info'].concat(args));
        this.hook(level, [msg].concat(args));
    };
    Logger.prototype.info = function (msg) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.log.apply(this, [msg, 'info'].concat(args));
    };
    Logger.prototype.error = function (msg) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.log.apply(this, [msg, 'error'].concat(args));
    };
    Logger.prototype.debug = function (msg) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.log.apply(this, [msg, 'debug'].concat(args));
    };
    Logger.prototype.warn = function (msg) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.log.apply(this, [msg, 'warn'].concat(args));
    };
    Logger.prototype.critical = function (msg) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.log.apply(this, [msg, 'critical'].concat(args));
    };
    Logger.prototype.hook = function (type, args) {
        if (this.senderHooks[type]) {
            this.senderHooks[type].forEach(function (hookAction) {
                hookAction.apply(void 0, args);
            });
        }
    };
    return Logger;
}());
exports.Logger = Logger;
var globalConfig = {};
var defaultConfig = {
    graylogHost: '127.0.0.1',
    graylogPort: 80,
    level: 'info'
};
var loggerCache = {};
var LoggerFactory = {
    globalConfig: function (config) {
        Object.assign(globalConfig, defaultConfig, config);
    },
    getLogger: function (logger, config) {
        if (!loggerCache[logger]) {
            loggerCache[logger] = new Logger(Object.assign({}, defaultConfig, globalConfig, config, { logger: logger }));
        }
        return loggerCache[logger];
    }
};
exports.default = LoggerFactory;
module.exports = LoggerFactory;
