// @td/node-shutter/lib/Properties.js
"use strict";
/// <reference path="../typings/index.d.ts" />
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var node_fetch_1 = require("node-fetch");
var crypto = require("crypto");
var Parser = require("properties");
var _ = require("lodash");
var path = require("path");
var fs = require("fs");
var constant_1 = require("./constant");
function noop() { }
var Properties = /** @class */ (function () {
    function Properties(options) {
        this.app = options.app;
        this.config = options.config;
        this.locations = options.locations;
        this.properties = {};
    }
    Properties.prototype.getRemoteUrl = function (location) {
        return "app=" + this.app + "&env=" + this.config.environment + "&group=" + this.config.group + "&key=" + location;
    };
    Properties.prototype.getProperty = function (key) {
        return this.properties[key];
    };
    Properties.prototype.setProperty = function (key, value) {
        this.properties[key] = value;
    };
    Properties.prototype.observer = function () {
        var _this = this;
        var hosts = this.config.hostList.split(',');
        var domain = "http://" + hosts[0] + "/v2/api/config/observer";
        var timestamp = Math.floor(new Date().getTime() / 1000);
        var requestBody = this.locations.map(function (location) { return ({
            app: _this.app,
            group: _this.config.group,
            key: location,
            hash: _this.hash,
            type: 0,
            ts: timestamp
        }); });
        return node_fetch_1.default(domain, {
            method: 'POST',
            body: JSON.stringify(requestBody),
            headers: __assign({ 'Content-Type': 'application/json' }, this.getQueryHeaders('', timestamp)),
            timeout: constant_1.OBSERVER_TIMEOUT
        })
            .then(function (res) { return res.json(); })
            .then(function (res) {
            if (res.success && res.data && res.data.length) {
                return _this.fetchData();
            }
        }).catch(function (e) {
            if (e.type !== 'request-timeout') {
                console.log('observer fetch error：', e);
            }
        }).then(function () {
            _this.watch(_this.watchHandle);
        });
    };
    Properties.prototype.watch = function (handle) {
        var _this = this;
        this.watchHandle = handle;
        setTimeout(function () {
            _this.observer();
        }, this.config.watchIntervalTime || 30000); // 默认30s请求一次
    };
    Properties.prototype.getCacheFilePath = function (location) {
        var app = this.app;
        var portalApplication = this.config.application;
        if (!fs.existsSync(constant_1.CACHE_DIR)) {
            fs.mkdir(constant_1.CACHE_DIR, noop);
        }
        var cacheFilePath = path.resolve(constant_1.CACHE_DIR, portalApplication + "~" + app + "~" + location);
        return cacheFilePath;
    };
    Properties.prototype.cacheFile = function (text, location) {
        var cacheFilePath = this.getCacheFilePath(location);
        fs.writeFile(cacheFilePath, text, noop);
    };
    Properties.prototype.readCacheFile = function (location) {
        var cacheFilePath = this.getCacheFilePath(location);
        if (fs.existsSync(cacheFilePath)) {
            return fs.readFileSync(cacheFilePath, 'utf8');
        }
        return '';
    };
    Properties.prototype.parseToAssign = function (text) {
        var _this = this;
        return new Promise(function (resolve) {
            Parser.parse(text, {}, function (err, obj) {
                console.log(text);
                if (err) {
                    console.log(err);
                    resolve(false);
                    return;
                }
                Object.assign(_this.properties, obj);
                resolve(true);
            });
        });
    };
    Properties.prototype.getQueryHeaders = function (plainText, timestamp) {
        var _a = this.config, AccessKey = _a.AccessKey, SecretKey = _a.SecretKey, application = _a.application;
        var ts = "" + timestamp + SecretKey;
        var stringToSign = !!plainText ? plainText + "&ts=" + ts : "ts=" + ts;
        var md5Hash = crypto.createHash('md5').update(stringToSign).digest('hex');
        return {
            'X-App-Id': application,
            'X-Access-Key': AccessKey,
            'X-Signature-Time': timestamp,
            'Authorization': md5Hash
        };
    };
    Properties.prototype.fetchData = function () {
        var _this = this;
        var promises = this.locations.map(function (location) {
            var timestamp = Math.floor(new Date().getTime() / 1000);
            var hosts = _this.config.hostList.split(',');
            var domain = "http://" + hosts[0] + "/v2/api/config/file";
            var remoteUrl = _this.getRemoteUrl(location);
            var queryHeaders = _this.getQueryHeaders(remoteUrl, timestamp);
            return node_fetch_1.default(domain + "?" + remoteUrl, {
                headers: queryHeaders
            })
                .then(function (res) {
                _this.hash = res.headers.get('X-Content-Hash');
                return res.buffer();
            })
                .then(function (buffer) { return buffer.toString(); })
                .then(function (res) { return new Promise(function (resolve, reject) {
                try {
                    var data = JSON.parse(res);
                    if (data && data.success === false) {
                        reject(new Error('get shutter file error'));
                        return;
                    }
                    resolve(res);
                }
                catch (e) {
                    resolve(res);
                }
            }); })
                .then(function (text) {
                _this.cacheFile(text, location);
                return _this.parseToAssign(text);
            }).catch(function () {
                var text = _this.readCacheFile(location);
                if (!text)
                    return false;
                return _this.parseToAssign(text);
            });
        });
        return Promise.all(promises)
            .then(function (results) { return results.findIndex(function (flag) { return !flag; }) !== -1; })
            .catch(function (e) { return console.log(e); });
    };
    Properties.prototype.getApp = function () {
        return this.app;
    };
    Properties.prototype.getLocations = function () {
        return this.locations;
    };
    Properties.prototype.addLocations = function (locations) {
        this.locations = _.uniq(this.locations.concat(locations));
    };
    return Properties;
}());
exports.Properties = Properties;
